csrc/scheduler/registry.cpp

```diff
@@ -2159,52 +2170,101 @@ class PersistentKernelScheduler : public SchedulerEntry {
     }
     const bool combined_inner_outer_reduction =
         inner_reduction_count && outer_reduction_count;
-    if (combined_inner_outer_reduction) {
-      persistent_buffer_size +=
-          normalization_scheduler_utils::partialReductionBufferSize(
-              outer_reduction_tvs, runtime_info);
-    }
 
-    // At this point, we use the full register file size only for the
-    // inner-outer case. It does not mean the full size shouldn't be used
-    // otherwise, but more detailed tuning of the heuristics would be required.
-    int64_t available_persistent_buffer_size = combined_inner_outer_reduction
-        ? scheduler_utils::register_file_size_full
+    // At this point, we use a much larger register file size for the combined
+    // case as it is rarely fused with other ops.
+    int64_t available_register_buffer_size = combined_inner_outer_reduction
+        ? scheduler_utils::register_file_size_combined
         : scheduler_utils::register_file_size;
 
-    // Use shared memory for persistent buffer is only tested for inner
-    // reduction
-    // TODO: extend to outer reduction and combined reduction
-    const bool allow_shared_memory =
-        inner_reduction_count > 0 && outer_reduction_count == 0;
-    if (allow_shared_memory) {
-      const auto dev_prop = at::cuda::getCurrentDeviceProperties();
-      const int64_t max_shared_memory_size =
-          (int64_t)dev_prop->sharedMemPerBlockOptin;
-      // Some shared memories are reserved for kernel launch overhead and
-      // reduction_broadcast_workspace. Estimation is conservative, but should
-      // be good enough. The actual threads per block is set in the heuristics
-      // and it may be smaller than maxThreadsPerBlock.
-      // TODO: More accurate estimation of available shared memory size.
-      const int64_t kernel_overhead =
-          (int64_t)dev_prop->reservedSharedMemPerBlock;
-      int64_t max_buffer_dtype_size = 1;
-      for (auto tv : persistent_buffer_info.persistent_buffers) {
-        max_buffer_dtype_size = std::max(
-            max_buffer_dtype_size,
-            dataTypeSize(
-                tv->getDataType().value(), runtime_info.getIndexType()));
-      }
-      const int64_t reduction_broadcast_workspace =
-          (int64_t)(dev_prop->maxThreadsPerBlock) * max_buffer_dtype_size;
-      const int64_t available_shared_memory_size = max_shared_memory_size -
-          kernel_overhead - reduction_broadcast_workspace;
-      available_persistent_buffer_size = std::max(
-          available_persistent_buffer_size, available_shared_memory_size);
-    }
+    // Some shared memories are reserved for kernel launch overhead and
+    // reduction_broadcast_workspace. Estimation is conservative, but should
```

Comment:
"Estimation is conservative, but should be good enough."? What do you mean by good enough?

Reply:
Excuse my English. What I was trying to say is :
```
  // The shared memory allocated for the reduction broadcast
  // workspace is proportional to the number of threads per block. Yet, since
  // the exact number of threads per block is unknown at this point,
  // maxThreadsPerBlock is used for a cautious estimate."
```